hw/mcu/sifive/fe310/src/hal_spi.c

```diff
@@ -0,0 +1,384 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+#include <os/mynewt.h>
+#include <assert.h>
+
+#include <hal/hal_spi.h>
+#include <hal/hal_gpio.h>
+
+#include <defs/error.h>
+#include <syscfg/syscfg.h>
+
+#include <mcu/fe310_hal.h>
+#include <mcu/sys_clock.h>
+#include <mcu/plic.h>
+#include <sifive/devices/spi.h>
+#include <env/freedom-e300-hifive1/platform.h>
+
+struct fe310_hal_spi
+{
+    struct hal_spi_settings spi_cfg;
+
+    uint32_t spi_base;
+
+    int len;
+    int txleft;
+    int rxleft;
+
+    /* Pointers to tx/rx buffers */
+    uint8_t *txbuf;
+    uint8_t *rxbuf;
+
+    /* Callback and arguments */
+    hal_spi_txrx_cb txrx_cb_func;
+    void            *txrx_cb_arg;
+};
+
+#if MYNEWT_VAL(SPI_1)
+struct fe310_hal_spi fe310_hal_spi1 = {
+    .spi_base = SPI1_CTRL_ADDR
+};
+#endif
+
+#if MYNEWT_VAL(SPI_2)
+struct fe310_hal_spi fe310_hal_spi2 = {
+    .spi_base = SPI2_CTRL_ADDR
+};
+#endif
+
+static struct fe310_hal_spi *fe310_hal_spis[3] = {
+    NULL, /* SPI 0 used to access flash */
```

Comment:
Is this always the case for this MCU?

Reply:
Mathilda: Is life always this hard, or is it just when you're a kid?
Léon: Always like this.

SPI0 is always for QSPI flash. Pins can't be configured as GPIO.
SPI2 is not routed to external Pins (in current hardwarefor now)
SPI1 is the only SPI that can be used.